frontend/pages/hosts/details/cards/Software/SelfService/SelfService.tsx

```diff
@@ -50,42 +60,296 @@ const SoftwareSelfService = ({
   pathname,
   queryParams,
   router,
+  onShowInstallerDetails,
 }: ISoftwareSelfServiceProps) => {
-  const { data, isLoading, isError, refetch } = useQuery<
-    IGetDeviceSoftwareResponse,
-    AxiosError,
-    IGetDeviceSoftwareResponse,
-    IDeviceSoftwareQueryKey[]
-  >(
-    [
+  const { renderFlash } = useContext(NotificationContext);
+
+  const pendingSoftwareSetRef = useRef<Set<string>>(new Set()); // Track for polling
+  const pollingTimeoutIdRef = useRef<NodeJS.Timeout | null>(null);
+
+  const queryKey = useMemo<IDeviceSoftwareQueryKey[]>(() => {
+    return [
       {
         scope: "device_software",
         id: deviceToken,
         page: queryParams.page,
+        query: queryParams.query,
         ...DEFAULT_SELF_SERVICE_QUERY_PARAMS,
       },
-    ],
-    ({ queryKey }) => deviceApi.getDeviceSoftware(queryKey[0]),
+    ];
+  }, [deviceToken, queryParams.page, queryParams.query]);
+
+  // Fetch self-service software (regular API call)
+  const {
+    data,
+    isLoading,
+    isError,
+    isFetching,
+    refetch: refetchSelfServiceSoftware,
+  } = useQuery<
+    IGetDeviceSoftwareResponse,
+    AxiosError,
+    IGetDeviceSoftwareResponse,
+    IDeviceSoftwareQueryKey[]
+  >(queryKey, (context) => deviceApi.getDeviceSoftware(context.queryKey[0]), {
+    ...DEFAULT_USE_QUERY_OPTIONS,
+    enabled: isSoftwareEnabled,
+    keepPreviousData: true,
+    staleTime: 7000,
+  });
+
+  // Poll for pending installs
+  const { refetch: refetchForPendingInstalls } = useQuery<
+    IGetDeviceSoftwareResponse,
+    AxiosError
+  >(
+    ["pending_installs", queryKey[0]],
+    () => deviceApi.getDeviceSoftware(queryKey[0]),
     {
-      ...DEFAULT_USE_QUERY_OPTIONS,
-      enabled: isSoftwareEnabled, // if software inventory is disabled, we don't bother fetching and always show the empty state
-      keepPreviousData: true,
-      staleTime: 7000,
+      enabled: false,
+      onSuccess: (response) => {
+        // Get the set of pending software IDs
+        const newPendingSet = new Set(
+          response.software
+            .filter((software) => software.status === "pending_install")
+            .map((software) => String(software.id))
+        );
+
+        // Compare new set with the previous set
+        const setsAreEqual =
+          newPendingSet.size === pendingSoftwareSetRef.current.size &&
+          [...newPendingSet].every((id) =>
+            pendingSoftwareSetRef.current.has(id)
+          );
+
+        if (newPendingSet.size > 0) {
+          // If the set changed, update and continue polling
+          if (!setsAreEqual) {
+            pendingSoftwareSetRef.current = newPendingSet;
+            refetchSelfServiceSoftware();
```

Comment:
This looks like we would fire off a request immediately if the set changed rather than waiting 5s? Seems like we're calling the same endpoint 2x in rapid succession? Which probably means I'm misunderstanding this logic.

Reply:
tyty